source/tokens.rs

```diff
@@ -702,15 +728,402 @@ named_attr!(
     )
 );
 
+/// A token is a structure pairing a span to any data.
+#[derive(Debug, PartialEq)]
+pub struct Token<'a, T> {
+    /// Value of the token.
+    pub value: T,
+
+    /// The attached span of the value.
+    pub span : Span<'a>
+}
+
+/// A span is a set of meta information about a token.
+///
+/// The `Span` structure can be used as an input of the nom parsers.
+#[derive(Debug, PartialEq, Copy, Clone)]
+pub struct Span<'a> {
+    /// The offset represents the position of the slice relatively to
+    /// the input of the parser. It starts at offset 0.
+    pub offset: usize,
+
+    /// The line number of the slice relatively to the input of the
+    /// parser. It starts at line 1.
+    pub line  : u32,
+
+    /// The column number of the slice relatively to the input of the
+    /// parser. It starts at column 1.
+    pub column: u16,
```

Comment:
OK, let's try.

For `usize`/`u32`/`u16` (resp. for `offset`, `line` and `column`):

```rust
struct Span<'a> {
    pub offset: usize,
    pub line: u32,
    pub column: u16,
    pub slice: &'a [u8]
}
```

We have this:

```sh
$ cargo rustc -- -Z print-type-sizes
…
print-type-size type: `Span`: 32 bytes, alignment: 8 bytes
print-type-size     field `.offset`: 8 bytes
print-type-size     field `.slice`: 16 bytes
print-type-size     field `.line`: 4 bytes
print-type-size     field `.column`: 2 bytes
print-type-size     end padding: 2 bytes
```

And for `usize`/`u32`/`u32` (so `column` has moved from `u16` to `u32`):

```rust
struct Span<'a> {
    pub offset: usize,
    pub line: u32,
    pub column: u32,
    pub slice: &'a [u8]
}
```

We have this:

```
$ cargo rustc -- -Z print-type-sizes
…
print-type-size type: `Span`: 32 bytes, alignment: 8 bytes
print-type-size     field `.offset`: 8 bytes
print-type-size     field `.slice`: 16 bytes
print-type-size     field `.line`: 4 bytes
print-type-size     field `.column`: 4 bytes
```

Size says the same, good alignment!

However, for `usize`/`usize`/`usize`, we have:

```sh
print-type-size type: `Span`: 40 bytes, alignment: 8 bytes
print-type-size     field `.offset`: 8 bytes
print-type-size     field `.line`: 8 bytes
print-type-size     field `.column`: 8 bytes
print-type-size     field `.slice`: 16 bytes
```

So it's heavier of 8 bytes.

In a tabular:

`offset` | `line` | `column` | total size of `Span`
---------|-------|-----------|--------------------
`usize`  | `u32` | `u16`      | 32 bytes
`usize` | `u32` | `u32`      | 32 bytes
`usize` | `usize` | `usize` | 40 bytes

So I guess we can move to `u32` for `column` with not a lot of fear :+1:!